### PR TITLE
Add Access Code checker

### PR DESCRIPTION
**Pull request by scouillard**
## Description

Update the `create_meeting_options` method in Room to check if the `glViewerAccessCode` and `glModeratorAccessCode `configs are enabled. If so, the upcoming created Rooms will have a generated access code.

## Testing Steps

- Enable or disable the access codes options in Admin/Room Configuration
- Create a new room
- Go to Room/Room Settings
- If enabled, the access code(s) should have been generated

### CODE REVIEW

**Comment by scouillard** [comment]
Tried to use the RoomSettingsGetter service, but it does not work with AccessCode because they are different than the other settings - they return an empty string or the access code instead of a boolean.

File: app/controllers/api/v1/rooms_controller.rb [hunk 1]
```diff
@@ -16,6 +16,8 @@ class RoomsController < ApiController
         ensure_authorized('ManageRooms', friendly_id: params[:friendly_id])
       end
 
+      after_action :access_code_checker, only: :create
```
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:19]: I'm not sure I like the `after_action` / `after_create`. I'd prefer if we move this logic directly into the `create` action instead

Comment (farhatahmad) [reply app/controllers/api/v1/rooms_controller.rb:19]: Since you're depending on the action initializing a `@new_room` for you

Comment (scouillard) [reply app/controllers/api/v1/rooms_controller.rb:19]: Yeah, well this would solve my current issue with Rspec. I'm having a hard time accessing the after_action in specs.

**Comment by scouillard** [comment]
Not super proud of the specs.
It is working but I would like to find a better solution for this.

File: app/controllers/api/v1/rooms_controller.rb [hunk 2]
```diff
@@ -106,6 +106,22 @@ def find_room
       def room_params
         params.require(:room).permit(:name, :user_id, :presentation)
       end
+
+      # Generates an access code on rooms#create if the access code room configuration is enabled
+      def access_code_checker
```
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:111]: I would call this `generate_access_code` or something like that

File: app/controllers/api/v1/rooms_controller.rb [hunk 3]
```diff
@@ -106,6 +106,22 @@ def find_room
...
+          room_id: @new_room.id,
+          provider: current_provider,
+          current_user:,
+          show_codes: false,
+          settings: %w[glViewerAccessCode glModeratorAccessCode]
+        ).call
+
+        access_code_settings.each do |access_code_setting|
+          room_meeting_option = RoomMeetingOption.joins(:meeting_option).where(room: @new_room, meeting_option: { name: access_code_setting })
+          room_meeting_option.update(value: SecureRandom.alphanumeric(6).downcase)
```
Comment (farhatahmad) [inline app/controllers/api/v1/rooms_controller.rb:122]: You can do this in one line 
```
RoomMeetingOption
  .joins(:meeting_option)
  .where(room: @new_room, meeting_option: { name: access_code_setting })
  .update(value: SecureRandom.alphanumeric(6).downcase)
```

File: spec/controllers/rooms_controller_spec.rb [hunk 4]
```diff
@@ -173,6 +173,26 @@
...
+        allow(fake_room_settings_getter).to receive(:call).and_return(
+          { 'glViewerAccessCode' => true, 'glModeratorAccessCode' => false }
+        )
+        # can we avoid hardcoding the friendly id?
+        allow(SecureRandom).to receive(:alphanumeric).and_return('8ac31fcc-b550-48f3-a1cd-eab7ec413f35', 'abc123')
+      end
+
+      it 'creates a room with pre-generated access codes' do
+        post :create, params: room_params
+        expect(Room.last.room_meeting_options[0].value).to eq('abc123') # there is a better way to access the room resource
```
Comment (farhatahmad) [inline spec/controllers/rooms_controller_spec.rb:192]: Just to be 100% safe, you should look up the Room based on the room_params (by user_id and room name)

File: spec/controllers/rooms_controller_spec.rb [hunk 5]
```diff
@@ -173,6 +173,26 @@
...
+      let!(:meeting_option) { create(:meeting_option, name: 'glViewerAccessCode') }
+      let!(:room_meeting_option) { create(:room_meeting_option, meeting_option:) }
+
+      before do
+        allow(RoomSettingsGetter).to receive(:new).and_return(fake_room_settings_getter)
+        allow(fake_room_settings_getter).to receive(:call).and_return(
+          { 'glViewerAccessCode' => true, 'glModeratorAccessCode' => false }
+        )
+        # can we avoid hardcoding the friendly id?
+        allow(SecureRandom).to receive(:alphanumeric).and_return('8ac31fcc-b550-48f3-a1cd-eab7ec413f35', 'abc123')
```
Comment (farhatahmad) [inline spec/controllers/rooms_controller_spec.rb:187]: You don't need this check - you can just check that it's not blank below

File: spec/controllers/rooms_controller_spec.rb [hunk 6]
```diff
@@ -173,6 +173,26 @@
       expect(response).to have_http_status(:forbidden)
     end
 
+    context 'the access code meeting option is enabled' do
+      let(:fake_room_settings_getter) { instance_double(RoomSettingsGetter) }
+      let!(:meeting_option) { create(:meeting_option, name: 'glViewerAccessCode') }
+      let!(:room_meeting_option) { create(:room_meeting_option, meeting_option:) }
```
Comment (farhatahmad) [inline spec/controllers/rooms_controller_spec.rb:179]: Why are these needed if you're stubbing the RoomSettingsGetter?

File: spec/controllers/rooms_controller_spec.rb (hunk 4 above)
Comment (scouillard) [reply spec/controllers/rooms_controller_spec.rb:192]: Great idea - thank you

File: spec/controllers/rooms_controller_spec.rb (hunk 5 above)
Comment (scouillard) [reply spec/controllers/rooms_controller_spec.rb:187]: The RoomMeetingOption factory is doing `value { [true, false, '12345'].sample }`
I will see if I can play with it without breaking the other specs.

File: spec/controllers/rooms_controller_spec.rb (hunk 6 above)
Comment (scouillard) [reply spec/controllers/rooms_controller_spec.rb:179]: The RoomSettingsGetter will return the settings values, but AFAIK, it doesn't actually create the RoomMeetingOption and MeetingOption? I will double check.

File: spec/controllers/rooms_controller_spec.rb (hunk 5 above)
Comment (farhatahmad) [reply spec/controllers/rooms_controller_spec.rb:187]: I'm not sure what you mean. The value in the db is not relevant since you're stubbing the RoomSettingsGetter response anyways.

Comment (scouillard) [reply spec/controllers/rooms_controller_spec.rb:187]: Maybe there is something I didn't get, but the RoomSettingsGetter returns a true or false depending if the access_code option is enabled or not. 

If I want to get the actual value of the access_code, the string value, I have to get it elsewhere.

In the specs, and in the room#generate_access_code, I will find the value by matching the RoomMeetingOption/MeetingOption with the room id and the name of the meeting option.

File: spec/controllers/rooms_controller_spec.rb (hunk 6 above)
Comment (farhatahmad) [reply spec/controllers/rooms_controller_spec.rb:179]: You shouldn't need to - the `generate_access_code` method should create the RoomMeetingOption for you

File: app/controllers/api/v1/rooms_controller.rb [hunk 7]
```diff
@@ -51,13 +51,13 @@ def public_show
...
-        # user_id = room_params[:user_id]
-        room = Room.create(name: room_params[:name], user_id: room_params[:user_id])
+        @new_room = Room.create(name: room_params[:name], user_id: room_params[:user_id])
 
         return render_error errors: user.errors.to_a if hcaptcha_enabled? && !verify_hcaptcha(response: params[:token])
 
-        if room.save
-          logger.info "room(friendly_id):#{room.friendly_id} created for user(id):#{room.user_id}"
+        if @new_room.save
+          generate_access_code
```
Comment (hadicheaito1) [inline app/controllers/api/v1/rooms_controller.rb:59]: Wouldn't it be better to just create a `room` and pass in that room to the `generate_access_code` function instead of creating a global var `new_room` ?

Comment (scouillard) [reply app/controllers/api/v1/rooms_controller.rb:59]: Yes, passing the room as an argument would be better.
I can't recall why I did it that way.
This whole PR is getting a big revamp soon - the logic is being put inside the Room model instead of in the controller.

File: app/controllers/api/v1/rooms_controller.rb [hunk 8]
```diff
@@ -51,13 +51,13 @@ def public_show
...
         # The created room will be the current user's unless a user_id param is provided with the request.
-        # user_id = room_params[:user_id]
-        room = Room.create(name: room_params[:name], user_id: room_params[:user_id])
+        @new_room = Room.create(name: room_params[:name], user_id: room_params[:user_id])
 
         return render_error errors: user.errors.to_a if hcaptcha_enabled? && !verify_hcaptcha(response: params[:token])
 
-        if room.save
-          logger.info "room(friendly_id):#{room.friendly_id} created for user(id):#{room.user_id}"
+        if @new_room.save
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/rooms_controller.rb:58]: The room should be already created,
I think a test like `persisted?` is more adequate.

Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:58]: If it's feasible using `create!`  or changing the logic to first instantiate a room model `new` and then `save` it  may help

File: app/controllers/api/v1/rooms_controller.rb [hunk 9]
```diff
@@ -106,6 +106,23 @@ def find_room
       def room_params
         params.require(:room).permit(:name, :user_id, :presentation)
       end
+
+      # Generates an access code on rooms#create if the access code room configuration is enabled
+      def generate_access_code
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/rooms_controller.rb:111]: I don't think that is indeed generating access codes if they are forced enabled on rooms creation,
I think it's auto generating access codes for created rooms all time,

File: app/controllers/api/v1/rooms_controller.rb [hunk 10]
```diff
@@ -106,6 +106,23 @@ def find_room
...
+        access_code_settings = RoomSettingsGetter.new(
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/rooms_controller.rb:112]: Besides, I believe this as is isn't doing anything except making an extra query

File: app/controllers/api/v1/rooms_controller.rb [hunk 11]
```diff
@@ -106,6 +106,23 @@ def find_room
...
+          room_id: @new_room.id,
+          provider: current_provider,
+          current_user:,
+          show_codes: false,
+          settings: %w[glViewerAccessCode glModeratorAccessCode]
+        ).call
+
+        access_code_settings.each do |access_code_setting|
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/rooms_controller.rb:120]: I believe that the `access_code_settings` will always be `%w[glViewerAccessCode  glModeratorAccessCode]`,

Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:120]: Besides, all points mentioned above.
Using `Hash#each_key` seems to be more adequate since not using the values,

**Review by KH-Amir-TN — CHANGES_REQUESTED** [review]
I think this is a model concern and can be further simplified and optimized if made as part of the `Room#create_meeting_options`.
We can generate access codes on the run as we create the access codes room meeting options, if the access code room config is forced enabled then we generate and set it on creation otherwise we default it.

**Comment by scouillard** [comment]
> I think this is a model concern and can be further simplified and optimized if made as part of the `Room#create_meeting_options`. We can generate access codes on the run as we create the access codes room meeting options, if the access code room config is forced enabled then we generate and set it on creation otherwise we default it.

Absolutely,

I learnt about create_meeting_options yesterday.
I am almost done revamping the PR, moving the logic from the controller to the model - I'm sorry I should have mentioned it.
I am learning from your PR reviews so it is not in vain.

File: app/controllers/api/v1/rooms_controller.rb (hunk 9 above)
Comment (KH-Amir-TN) [reply app/controllers/api/v1/rooms_controller.rb:111]: That's odd, I tested either.
Maybe, I missed something.

File: app/models/room.rb [hunk 12]
```diff
@@ -37,6 +36,21 @@ def get_setting(name:)
                         .find_by(meeting_option: { name: })
   end
 
+  # Autocreate all meeting options using the default values
+  def create_meeting_options(provider)
+    rooms_configs = MeetingOption.joins(:rooms_configurations)
```
Comment (KH-Amir-TN) [inline app/models/room.rb:41]: I think if we extend `MeetingOption#get_config_value` to fetch single/multiple room configs by name and provider will reduce this boilerplate code and we can leverage it throughout the application.

File: app/controllers/api/v1/rooms_controller.rb [hunk 13]
```diff
@@ -51,12 +51,12 @@ def public_show
       def create
         # TODO: amir - ensure accessibility for authenticated requests only.
         # The created room will be the current user's unless a user_id param is provided with the request.
-        # user_id = room_params[:user_id]
         room = Room.create(name: room_params[:name], user_id: room_params[:user_id])
 
         return render_error errors: user.errors.to_a if hcaptcha_enabled? && !verify_hcaptcha(response: params[:token])
 
         if room.save
+          room.create_meeting_options(current_provider)
```
[the remark this answers is not in the document]
Comment (scouillard) [reply app/controllers/api/v1/rooms_controller.rb:59]: I couldn't find a solution to access the Provider inside the model

File: spec/factories/meeting_option_factory.rb [hunk 14]
```diff
@@ -2,7 +2,7 @@
 
 FactoryBot.define do
   factory :meeting_option do
-    name { Faker::Vehicle.unique.make }
+    name { Faker::Address.unique.street_address }
```
[the remark this answers is not in the document]
Comment (scouillard) [reply spec/factories/meeting_option_factory.rb:5]: Faker was throwing an error.
There is not enough `Vehicule.make`, which is a problem because of the unique tag.
I double checked and it doesn't seem like we are creating a crazy amount of meeting_options.

File: app/models/room.rb (hunk 12 above)
Comment (scouillard) [reply app/models/room.rb:41]: I'll work on this today

File: app/models/room.rb [hunk 15]
```diff
@@ -37,6 +36,21 @@ def get_setting(name:)
                         .find_by(meeting_option: { name: })
   end
 
+  # Autocreate all meeting options using the default values
+  def create_meeting_options(provider)
```
[the remark this answers is not in the document]
Comment (scouillard) [reply app/models/room.rb:40]: I didn't think about using the room owner provider, good call.

File: spec/factories/meeting_option_factory.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (scouillard) [reply spec/factories/meeting_option_factory.rb:5]: @KH-Amir-TN has suggested to clear the cache.
It seems like the uniques stored in cache does not get cleared in between spec groups.
`Vehicle.unique.make` is also used in `room_settings_getter_spec.rb`

File: app/models/room.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (scouillard) [reply app/models/room.rb:40]: Thank you Amir for the great solution!

File: app/models/room.rb (hunk 12 above)
Comment (scouillard) [reply app/models/room.rb:41]: Done! `get_config_value` returns a hash with the configs values.

File: spec/factories/meeting_option_factory.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (scouillard) [reply spec/factories/meeting_option_factory.rb]: I think using Lorem.word is more adequate than Vehicle.make and should solve are our unique scarcity issues.

File: spec/controllers/rooms_controller_spec.rb [hunk 16]
```diff
@@ -173,6 +173,11 @@
       expect(response).to have_http_status(:forbidden)
     end
 
+    it 'calls create_meeting_options on room' do
```
Comment (KH-Amir-TN) [inline spec/controllers/rooms_controller_spec.rb:176]: I think we can omit this if we're testing the callback in the room model spec.

File: app/controllers/api/v1/room_settings_controller.rb [hunk 17]
```diff
@@ -15,15 +15,17 @@ def show
...
-        config = MeetingOption.get_config_value(name: room_setting_params[:settingName], provider: current_provider)&.value
-        return render_error status: :bad_request unless config
-
         name = room_setting_params[:settingName]
         value = room_setting_params[:settingValue].to_s
 
+        config = MeetingOption.get_config_value(name:, provider: current_provider)
+        config_value = config[name]
+
+        return render_error status: :bad_request unless config_value
```
Comment (KH-Amir-TN) [inline app/controllers/api/v1/room_settings_controller.rb:24]: If we're adding this then I think we may need to add a test for this in the `room_settings_controller_spec`.

File: spec/controllers/room_settings_controller_spec.rb [hunk 18]
```diff
@@ -41,6 +41,7 @@
 
   describe '#update' do
     it 'uses MeetingOption::get_config_value and updates the setting if its config is "optional"' do
+      expect(MeetingOption).to receive(:get_config_value).and_call_original
```
Comment (KH-Amir-TN) [inline spec/controllers/room_settings_controller_spec.rb:44]: Is this required?

File: spec/models/meeting_option_spec.rb [hunk 19]
```diff
@@ -9,7 +9,7 @@
     it { is_expected.to validate_uniqueness_of(:name) }
   end
 
-  describe 'classs methods' do
+  describe 'class methods' do
```
Comment (KH-Amir-TN) [inline spec/models/meeting_option_spec.rb:12]: My bad xD

File: spec/models/meeting_option_spec.rb [hunk 20]
```diff
@@ -40,15 +40,15 @@
           create(:rooms_configuration, meeting_option:, provider: 'greenlight', value: 'optional')
 
           rooms_config = described_class.get_config_value(name: 'setting', provider: 'greenlight')
-          expect(rooms_config.value).to eq('optional')
+          expect(rooms_config['setting']).to eq('optional')
```
Comment (KH-Amir-TN) [inline spec/models/meeting_option_spec.rb:43]: Um, nice!
We may even test this multi configs fetching capability?

File: spec/models/room_spec.rb [hunk 21]
```diff
@@ -54,12 +54,41 @@
     end
   end
 
-  describe 'after_create' do
```
Comment (KH-Amir-TN) [inline spec/models/room_spec.rb:57]: Why we're changing the structure?

File: spec/models/meeting_option_spec.rb (hunk 20 above)
Comment (KH-Amir-TN) [inline spec/models/meeting_option_spec.rb:43]: Um, nice!
We may even test this multi configs fetching capability?

File: spec/models/room_spec.rb [hunk 22]
```diff
@@ -54,12 +54,41 @@
...
-        expect { create(:room) }.to change(RoomMeetingOption, :count).from(0).to(5)
+    before do
+      create_list(:meeting_option, 5)
+    end
+
+    it 'creates a RoomMeetingOption for each MeetingOption' do
+      expect { create(:room) }.to change(RoomMeetingOption, :count).from(0).to(6)
+    end
+
+    it 'does not generate an access code if the room config is not enabled' do
```
Comment (KH-Amir-TN) [inline spec/models/room_spec.rb:68]: Access code feels like a test group, we can group those examples under a `context` maybe?

File: spec/controllers/room_settings_controller_spec.rb (hunk 18 above)
Comment (scouillard) [reply spec/controllers/room_settings_controller_spec.rb:44]: It is required because `get_config_value` is called in the `after_create create_meeting_options` callback in Room.

File: spec/models/meeting_option_spec.rb (hunk 20 above)
Comment (scouillard) [reply spec/models/meeting_option_spec.rb:43]: Isn't it being tested already since we are returning multi configs in `create_meeting_options` and a single config in `room_settings#update` ?

File: spec/models/room_spec.rb (hunk 21 above)
Comment (scouillard) [reply spec/models/room_spec.rb:57]: Ah good catch, it's because we removed the `after_create callback` and then we put it back.

File: app/controllers/api/v1/room_settings_controller.rb (hunk 17 above)
Comment (scouillard) [reply app/controllers/api/v1/room_settings_controller.rb:24]: I'm not sure what you want to test

File: spec/models/room_spec.rb (hunk 22 above)
Comment (scouillard) [reply spec/models/room_spec.rb:68]: I put the after_create context, and upon review I dont think an extra context is necessary

File: app/controllers/api/v1/room_settings_controller.rb (hunk 17 above)
Comment (KH-Amir-TN) [reply app/controllers/api/v1/room_settings_controller.rb:24]: The case where the action will return a `bad_request` response to the client.
Like with the unfound 404 tests.

File: spec/controllers/room_settings_controller_spec.rb (hunk 18 above)
Comment (KH-Amir-TN) [reply spec/controllers/room_settings_controller_spec.rb:44]: Um, I see what you mean.
However, I think we can further clean it.
If you intentionally wanna ensure that the method gets called twice you can assert that more specifically rather being so generic (that brakes a bit the resiliency of the tests).
Example instead of:
` expect(MeetingOption).to receive(:get_config_value).and_call_original`
This is more specific:
` expect(MeetingOption).to receive(:get_config_value).with(name: %w[glViewerAccessCode glModeratorAccessCode], provider: 'greenlight'| room.user.provider).and_call_original`
That ensures that if called that should be coming from the `Room#create_meeting_options`.
You can also stub the `Room#create_meeting_options` itself to isolate your test if it's implementation details are not relevant to this test. 
There's also a lot of what you can leverage from the Rspec mocking API that can better answer your need for this matter,

File: spec/models/meeting_option_spec.rb (hunk 20 above)
Comment (KH-Amir-TN) [reply spec/models/meeting_option_spec.rb:43]: Are we testing that in the `create_meeting_options` tests?
However, that's an integration test I think we may like to initially target that in its unit test catching all cases,
In example, in the `create_meeting_options` test you were only creating a `glViewerAccessCode` meeting option record only so to ensure that the multi configs fetching is working is only because we know the implementation of the actual method,

File: app/controllers/api/v1/room_settings_controller.rb (hunk 17 above)
Comment (scouillard) [reply app/controllers/api/v1/room_settings_controller.rb:24]: Seems to be unrelated to this PR

**Review by KH-Amir-TN — APPROVED** [review]

**Review by KH-Amir-TN — APPROVED** [review]

**Review by farhatahmad — APPROVED** [review]
